### PR TITLE
Strided masked reductions: sum, prod, amin, amax

### PR DESCRIPTION
**Pull request by pearu**
This PR implements a prototype of masked reductions according to https://github.com/pytorch/rfcs/pull/27 specifications.

- [x] Masked reductions docstring generation
- [ ] masked_sum
  * [ ] reference implementation
    + [x] strided
    + [x] sparse_coo
    + [ ] sparse_csr
  * [x] ReductionOpInfo
  * [ ] optimized implementation for strided input/mask
    + [ ] CPU
    + [ ] CUDA
    + [ ] benchmarks: `masked_sum(input, mask)` vs `sum(input * mask)`
  * [ ] optimized implementation for sparse input/mask
- [ ] masked_prod
  * [ ] reference implementation
    + [x] strided
  * [x] ReductionOpInfo
- [ ] masked_mean
- [ ] masked_amin
  * [ ] reference implementation
    + [x] strided
  * [x] ReductionOpInfo
- [ ] masked_amax
  * [ ] reference implementation
    + [x] strided
  * [x] ReductionOpInfo

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @heitorschueroff

### CODE REVIEW

File: torch/sparse/__init__.py [hunk 1]
```diff
@@ -201,3 +201,154 @@ def log_softmax(input: Tensor, dim: int, dtype: Optional[DType] = None) -> Tenso
...
+{masked_reduction_example}
+    >>> torch.sparse.masked_sum(input, mask, 1)
+    tensor([-4,  0])
+    """
+    # TODO: What follows is a reference implementation of masked sum
+    # that is to be replaced with an optimized one.
+    if mask is None:
+        input_mask = torch.as_tensor(input)
+    else:
+        input_mask = torch.as_tensor(input) * torch.as_tensor(mask)
```
Comment (cpuhrsch) [inline torch/sparse/__init__.py:321]: You actually have to use masked_fill because
```
>>> torch.tensor([float('inf')]) * torch.tensor([False])
tensor([nan])
>>> torch.tensor([float('inf')]).masked_fill(~torch.tensor([False]), 0)
tensor([0.])
```
will cause NaNs in the reduction.

Also, sounds like a good test case to feed NaNs and Infs into masked out elements and masking sure they don't propagate

File: torch/sparse/__init__.py [hunk 2]
```diff
@@ -201,3 +201,154 @@ def log_softmax(input: Tensor, dim: int, dtype: Optional[DType] = None) -> Tenso
...
+
+    # Expose function as public symbol
+    __all__.append(func.__name__)
+
+    return func
+
+
+@_apply_docstring_templates
+def masked_sum(input: Tensor,
+               mask: Optional[Tensor] = None,
```
Comment (cpuhrsch) [inline torch/sparse/__init__.py:299]: Why not add it to the end? That way it'll be a replacement positionally as well (I think)

File: torch/sparse/__init__.py [hunk 3]
```diff
@@ -201,3 +201,154 @@ def log_softmax(input: Tensor, dim: int, dtype: Optional[DType] = None) -> Tenso
...
+    >>> torch.sparse.masked_prod(input, mask, 1)
+    tensor([3,  1])
+    """
+    # TODO: What follows is a reference implementation of masked
+    # product that is to be replaced with an optimized one.
+    if mask is None:
+        input_mask = torch.as_tensor(input)
+    else:
+        mask = torch.as_tensor(mask)
+        input_mask = torch.as_tensor(input) * mask + ~mask
```
Comment (cpuhrsch) [inline torch/sparse/__init__.py:351]: same as above

File: torch/sparse/__init__.py (hunk 1 above)
Comment (cpuhrsch) [reply torch/sparse/__init__.py:321]: I'm also wondering whether it's necessary to explicitly cast mask to the dtype of the input Tensor. What if the input is bool or int8?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -837,6 +837,56 @@ def sample_inputs_reduction(op_info, device, dtype, requires_grad, **kwargs):
...
+            inputs.append(SampleInput(sample_input.input,
+                                      args=(mask,) + sample_input.args,
+                                      kwargs=sample_input.kwargs))
+    return inputs
+
+
+def reference_masked_sum(input, mask, dim=None, keepdim=False, dtype=None):
+    """Reference implementation of masked_sum.
+
+    See RFC-0016 for more information.
```
Comment (cpuhrsch) [inline torch/testing/_internal/common_methods_invocations.py:860]: For convenience and forward compatibility it could help to add the explicit link ```https://github.com/pytorch/rfcs/pull/27```

File: test/test_reductions.py [hunk 5]
```diff
@@ -2992,6 +2999,46 @@ def test_reduction_empty_any_all(self, device):
             self.assertEqual(torch.ones((2, 1, 4), device=device, dtype=out_dtype), xb.all(1, keepdim=True))
             self.assertEqual(torch.ones((), device=device, dtype=out_dtype), xb.all())
 
+    @ops(reference_masked_ops)
+    def test_reference_masked(self, device, dtype, op):
+        # Typically, op.ref is a reference function using numpy
```
Comment (cpuhrsch) [inline test/test_reductions.py:3004]: I think our semantics agree with [MaskedArray](https://numpy.org/doc/stable/reference/generated/numpy.ma.sum.html) though (aside from the inverted mask), right?

Comment (cpuhrsch) [reply test/test_reductions.py:3004]: I'm also wondering whether it makes sense to add an autograd test, in particular around things like a fully masked out row in a 2-D Tensor. What will be the gradient? This will matter once we write one-off kernels and there might be some subtleties to consider here.

File: torch/sparse/__init__.py (hunk 1 above)
Comment (pearu) [reply torch/sparse/__init__.py:321]: @cpuhrsch This implementation is not final nor correct as noted in the TODO comment and what you observed. This implementation should be ignored at this stage of the PR, in fact, the PR was created to discuss the testing approach as it uses OpInfo support in an untypical way (reference is pytorch implementation rather than numpy one).

File: torch/sparse/__init__.py (hunk 2 above)
Comment (pearu) [reply torch/sparse/__init__.py:299]: The answer depends on how tied are we to the definition of reduction operations by OpInfo:
https://github.com/pytorch/pytorch/blob/5060b69d62943e11f053f8c86b8154a7d009af11/torch/testing/_internal/common_methods_invocations.py#L860-L865
and how to interpret the usage of `*args` and `**kwargs` here.
I don't have strong preferences but it seemed to me that putting `mask` to `*args` part would be a first natural choice.
Following https://numpy.org/doc/stable/reference/generated/numpy.sum.html?highlight=sum#numpy.sum, the end position would be more appropriate indeed.

Comment (cpuhrsch) [reply torch/sparse/__init__.py:299]: To me it's a keyword argument in the sense that it's optional. By default it is ignored. I suggest we get another person involved here to see how to best make a decision after we've revised the other parts of this PR.

File: test/test_reductions.py (hunk 5 above)
Comment (pearu) [reply test/test_reductions.py:3004]: > I think our semantics agree with [MaskedArray](https://numpy.org/doc/stable/reference/generated/numpy.ma.sum.html) though (aside from the inverted mask), right?

Yes, I think this is true and my current plan is to use numpy.ma tools for reference as implementing the reference from scratch is not trivial considering possible nan/inf inputs.

Comment (pearu) [reply test/test_reductions.py]: Now testing masked reductions on strided tensors use numpy reduction operations.

File: torch/sparse/__init__.py (hunk 1 above)
Comment (pearu) [reply torch/sparse/__init__.py]: Added tests with nan/inf and replaced input_mask computation to use `where`.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py]: reference_masked_sum is removed.

**Comment by cpuhrsch** [comment]
Just saw that you're also already including support for sparse layouts in this PR. I think it might be easier to get started if we just merge the dense variant first and then start expanding support for sparse layouts. That way we can first nail the semantics for both forward and backward and then introduce sparsity as an optimizaiton.

File: torch/sparse/__init__.py (hunk 3 above)
Comment (pearu) [reply torch/sparse/__init__.py]: Replaced the implementation.

File: torch/sparse/__init__.py [hunk 6]
```diff
@@ -201,3 +201,269 @@ def log_softmax(input: Tensor, dim: int, dtype: Optional[DType] = None) -> Tenso
...
+            return torch.prod(mask_input, dim, keepdim, dtype=dtype)
+        else:
+            raise NotImplementedError(f'masked_prod of {input.layout} tensor (shape={input.shape}) for non-integer dim={dim}')
+    # TODO: elif input.layout == torch.sparse_coo:
+    # TODO: elif input.layout == torch.sparse_csr:
+    else:
+        raise NotImplementedError(f'masked_prod of {input.layout} tensor')
+
+
+def masked_mask(input: Tensor,
```
Comment (cpuhrsch) [inline torch/sparse/__init__.py:430]: Can't we use any for this?

Comment (pearu) [reply torch/sparse/__init__.py:430]: `torch.any` is restricted to integer `dim`.
Until `torch.any` is generalized, we'll need `masked_mask` for testing masked reduction operations, specifically, for non-integral `dim` and broadcasting mask support.

Comment (cpuhrsch) [reply torch/sparse/__init__.py:430]: Could I ask you to add a leading underscore then or maybe move it into the tests only? We can then add an issue that requests multiple dimension support for torch.any. Seems like a reasonable feature request and we don't have to add a function that does the same thing, but with slightly more general functionality. Could you submit a feature request to add support for multiple dim for torch.any if you agree?

Comment (pearu) [reply torch/sparse/__init__.py:430]: An issue for supporting multiple dim in `torch.any` already exists:  https://github.com/pytorch/pytorch/issues/56586
I am ok with renaming `masked_mask` to `_masked_mask`.
Since output mask is needed for masked-equality-check in several test files and it also handles broadcasting to input shape (that `torch.any` would not do), I would leave `_masked_mask` in `torch.sparse` unless a better place is suggested.

Comment (cpuhrsch) [reply torch/sparse/__init__.py]: Sounds good. Given that we have an official tracking issue and multidim support for our masked reductions is needed for feature parity, it might make sense for us to send a diff that addresses the multidim support for torch.any Given that there's a tracking issue it should be readily accepted. https://github.com/pytorch/pytorch/issues/61582

**Comment by pearu** [comment]
https://github.com/pytorch/pytorch/pull/65990 will supersede this PR.
